lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,213 @@
+    type: str
+    returned: on failure
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def clean_err(err):
+    # Remove whitespaces and newlines that the snap CLI outputs.
+    return err.replace('\n', '').replace('      ', '').replace('"', '\'')
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'info', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'list', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc == 0
+
+
+def install_snaps(module, snap_names):
+    snaps_not_installed = list()
+    for snap_name in snap_names:
+        if is_snap_installed(module, snap_name):
+            continue
+        snaps_not_installed.append(snap_name)
+
+    if not snaps_not_installed:
+        snaps_already_installed = ', '.join(snap_names)
+        module.exit_json(msg="Snap(s) already installed: %s" % snaps_already_installed, classic=module.params['classic'], changed=False)
+
+    # Transform the list into a string with whitespace-separated snaps
+    snaps_to_install = ' '.join(snaps_not_installed)
+    # Create a string with commas for the output
+    snaps_installed = ', '.join(snaps_not_installed)
+
+    if module.check_mode:
+        module.exit_json(msg="Snap(s) that would have been installed: %s" % snaps_installed, classic=module.params['classic'], changed=True)
+
+    classic = '--classic' if module.params['classic'] else ''
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'install', snaps_to_install, classic]
+    cmd = ' '.join(cmd_parts)
+
+    # Actually install the snaps
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        module.exit_json(msg="Snap(s) installed: %s" % snaps_installed, classic=module.params['classic'], changed=True)
+    else:
+        err = clean_err(err)
+        module.fail_json(msg="Something went wrong.", classic=module.params['classic'], err=err)
+
+
+def remove_snaps(module, snap_names):
+    snaps_installed = list()
+    for snap_name in snap_names:
+        if is_snap_installed(module, snap_name):
+            continue
+        snaps_installed.append(snap_name)
+    if not snaps_installed:
+        snaps_not_installed = ', '.join(snap_names)
+        module.exit_json(msg="Snap(s) not installed: %s" % snaps_not_installed, changed=False)
+
+    # Transform the list into a string with whitespace-separated snaps
+    snaps_to_remove = ' '.join(snaps_installed)
+    # Create a string with commas for the output
+    snaps_removed = ', '.join(snaps_installed)
+
+    if module.check_mode:
+        module.exit_json(msg="Snap(s) that would have been removed: %s" % snaps_removed, changed=True)
```

Comment:
It's better to supply the actual list of snaps, not a string and via a separate var. This way they'll be machine-readable, facilitating users of the module to use them when they register result and want to process it somehow.

Reply:
So you propose to output `snaps_installed` directly?

Reply:
yep, this way someone can:
```yaml
- name: Install snaps
  snap:
    name:
    - vim
    - emacs
  register: install_result
- name: Packages installed
  debug: var={{ install_result.snaps_installed }}
- name: Check on Vim
  shell: ... # put vimrc in place
  when: >
    'vim' in install_result.snaps_installed
```

Reply:
I see... So all my `msg` should only output a list?

Reply:
Honestly, I don't see a point in putting some message there at all. `failed`/`changed` flags already indicate an overall status. `stdin`/`stdout` (and their `_lines` versions) are available to use automatically, so they can use that for result introspection.
Module's output should be useful robot-readable/parseable data, printing that to screen is done by callback plugins, if they want to print out a list, they'll probably do so based on raw data structure rather then parsing your message (it's more friendly API to give away lists/dicts, than an oddly formatted string with mashed data).

However, it might be useful to add information about what command this module executed to get job done or what were the module arguments (name (list), classic (bool?)). Look at this example `pip`'s return value:
```python
hostname | SUCCESS => {
    "changed": true, 
    "cmd": "/usr/bin/pip2 install -U tarsnapper", 
    "name": [
        "tarsnapper"
    ], 
    "requirements": null, 
    "state": "latest", 
    "stderr": "You are using pip version 8.1.1, however version 9.0.1 is available.\nYou should consider upgrading via the 'pip install --upgrade pip' command.\n", 
    "stderr_lines": [
        "You are using pip version 8.1.1, however version 9.0.1 is available.", 
        "You should consider upgrading via the 'pip install --upgrade pip' command."
    ], 
    "stdout": "Collecting tarsnapper\nCollecting argparse>=1.1 (from tarsnapper)\n  Using cached argparse-1.4.0-py2.py3-none-any.whl\nCollecting python-dateutil>=2.4.0 (from tarsnapper)\n  Using cached python_dateutil-2.6.1-py2.py3-none-any.whl\nCollecting pyyaml>=3.09 (from tarsnapper)\nCollecting pexpect>=3.1 (from tarsnapper)\n  Using cached pexpect-4.2.1-py2.py3-none-any.whl\nCollecting six>=1.5 (from python-dateutil>=2.4.0->tarsnapper)\n  Using cached six-1.10.0-py2.py3-none-any.whl\nCollecting ptyprocess>=0.5 (from pexpect>=3.1->tarsnapper)\n  Using cached ptyprocess-0.5.2-py2.py3-none-any.whl\nInstalling collected packages: argparse, six, python-dateutil, pyyaml, ptyprocess, pexpect, tarsnapper\nSuccessfully installed argparse-1.2.1 pexpect-4.2.1 ptyprocess-0.5.2 python-dateutil-2.6.1 pyyaml-3.12 six-1.10.0 tarsnapper-0.4.0\n", 
    "stdout_lines": [
        "Collecting tarsnapper", 
        "Collecting argparse>=1.1 (from tarsnapper)", 
        "  Using cached argparse-1.4.0-py2.py3-none-any.whl", 
        "Collecting python-dateutil>=2.4.0 (from tarsnapper)", 
        "  Using cached python_dateutil-2.6.1-py2.py3-none-any.whl", 
        "Collecting pyyaml>=3.09 (from tarsnapper)", 
        "Collecting pexpect>=3.1 (from tarsnapper)", 
        "  Using cached pexpect-4.2.1-py2.py3-none-any.whl", 
        "Collecting six>=1.5 (from python-dateutil>=2.4.0->tarsnapper)", 
        "  Using cached six-1.10.0-py2.py3-none-any.whl", 
        "Collecting ptyprocess>=0.5 (from pexpect>=3.1->tarsnapper)", 
        "  Using cached ptyprocess-0.5.2-py2.py3-none-any.whl", 
        "Installing collected packages: argparse, six, python-dateutil, pyyaml, ptyprocess, pexpect, tarsnapper", 
        "Successfully installed argparse-1.2.1 pexpect-4.2.1 ptyprocess-0.5.2 python-dateutil-2.6.1 pyyaml-3.12 six-1.10.0 tarsnapper-0.4.0"
    ], 
    "version": null, 
    "virtualenv": null
}
```

Note that `stderr`, `stderr_lines`, `stdout` and `stdout_lines` in this example have been generated *automatically*, so there's no real need to duplicate that in manually created keys with the same data. It's always available regardless.

Reply:
Oh, wait. I might be wrong about stdout/stderr. Need to test smth.

Reply:
Ok thanks. I had trouble finding what the convention is

Reply:
What should I return if I don't execute anything, and thus have no stdout/stderr?

Reply:
so don't return them in this case

Reply:
ok

Reply:
Reading the pip module, I assume I don't have to specify stdout/stderr in the RETURN schema